### ISSUE TITLE
Mutable torrents (BEP46)

### ISSUE THREAD

**Issue by lmatteis** [feross/webtorrent#886]
Hi guys, so I was thinking of implementing [BEP46](http://www.bittorrent.org/beps/bep_0046.html) in WebTorrent. I already implemented a simple reference implementation *on top* of it `webtorrent-cli`, but I think this needs to go deeper. So after couple hours of studying the code, I think the best place to put this should be in the `torrent-discovery`. It already has an interval (using recursive timeouts) which announces to the DHT (the regular announcing), so it makes sense to also have an interval that gets()/puts() if there's a mutable torrent.

Higher up in the hierarchy the `.add()` method of WebTorrent creates a `new Torrent()` EventEmitter, which emits `_infoHash`. This could instead emit the same event every time a new torrent is found (hence should simply change from `.once` to `.on`). Problem is whether we want to add a new torrent to the list, or should we update the current one?

Then inside `torrent.js` (always WebTorrent) the `new Discovery()` EventEmitter should probably emit an event for when a new torrent is found in the DHT. Something like `.on('update')` or something else more inline. 

These seem like most of the parts I'd have to touch. Thoughts @feross @substack @mafintosh ?

**Comment by feross** [comment]
Nice! The great thing about implementing this here instead of in your fork of `webtorrent-cli` is that all consumers of this library, like [WebTorrent Desktop](https://webtorrent.io/desktop), will inherit BEP46 support for free.

> Problem is whether we want to add a new torrent to the list, or should we update the current one when an update is found?

When a mutable torrent gets updated, the info hash of the torrent will change. That will violate a lot of assumptions about how `torrent.infoHash` works. It never changes right now and apps like WebTorrent Desktop will actually store the value and look up torrents later using `client.get(infoHash)`. Emitting an event and letting the user decide whether to update to the newer info hash could work.

Updating all the places where the info hash is used in a live torrent will be hard. It's used in lots of places and this sounds hard to get right.

Maybe the simplest way to implement this is in stages:

- Stage 1: Resolve mutable torrent magnet links. Treat `magnet:?xs=urn:btpk:[ Public Key (Hex) ]&s=[ Salt (Hex) ]` magnet links as "remote torrents" that need to be resolved before their info hash is known (like http links or filesystem paths to torrent files work today). Once the info hash is resolved via the DHT, then it's basically a normal torrent without updating.

- Stage 2: Make the torrents updatable. Periodically poll for changes to the info has and emit an event `torrent.on('mutableUpdate')` to let the user know that there's a new info hash. They can switch to the new info hash explicitly by calling `torrent.destroy()` and `client.add(torrent.magnetURI)` to add the new updated torrent.

- Stage 3: Find a higher-level API that's easier to use.

And some questions:

Question 1: Should data that's already on disk be re-used if the same file names are present in the new torrent? This might be impossible because if data is added to the middle of the torrent then all the piece hashes afterwards will be completely different.

Question 2: Should the old torrent immediately stop being shared?

**Comment by lmatteis** [comment]
Sounds good. Also to consider BEP39 (centralized version of BEP46) where updates of torrents happen through an HTTP server, so these APIs should be the same for this functionality as well. So it makes sense to have `torrent.on('update')` (for both BEP 39 and 46) and then call destroy() and add() accordingly. 

I'm a little bit ignorant about re-using the data. Can't I just run the new torrent on top of the old data so it automatically checks whether it can be reused? If the reuse routine errors out, we could replace the whole data.

**Comment by feross** [comment]
> So it makes sense to have torrent.on('update') (for both BEP 39 and 46) and then call destroy() and add() accordingly.

Supporting BEP39 as well sounds good to me.

> I'm a little bit ignorant about re-using the data.

If you specify the same path as the old torrent, and the new torrent has changed the filenames, then the old files will not be touched. This could confuse the user. For now, let's just leave this up to the user. They can delete the files if they want when they call `torrent.destroy()`.

**Comment by lmatteis** [comment]
I was thinking that most of the time publishers will apply updates to the underlying data their sharing. So re-downloading the whole data would be bad. For instance, imagine Archive.org publishing data dumps. As a user I wouldn't want to delete and replace the whole dump every time.  Ideas how http://www.bittorrent.org/beps/bep_0038.html can help?

**Comment by feross** [comment]
> Does webtorrent have capabilities to re-use data if I only appended data to my torrent?

I'm not sure how the `fs-chunk-store` package (which handles mapping the store of pieces to files on disk) will handle the situation where you point it to an existing file on disk, but the size has changed (gotten larger or smaller).

We should ensure that it handles this situation gracefully. If the file name is the same, but it's just larger now, then the file size should be increased on disk without deleting the file. Then, the normal piece verification process can take care of figuring out whether the data is valid or not.

But as I mentioned before, if the user specifies the same path as the old torrent but the new torrent has changed the filenames, then the old files will not be touched. That will leave files from the old torrent and new one side-by-side.

The only way to make data re-use work nicely for the user is if we handle this in WebTorrent. We need to add say a `torrent.applyUpdate()` method which will switch to the new info hash, and delete whichever files are no longer present in the new torrent.

If you want to tackle this in the same PR, you can give it a go. But I think it would be easier to just punt on this for now and add it in a future PR.

As an initial PR, I would just fire an event and allow the user to call `torrent.destroy()` and `client.add(torrent.magnetURI)` manually. Does this seem reasonable to you?

**Comment by lmatteis** [comment]
@feross so the first step is to parse the new magnet URI structure. I've nailed it down to this part of code:

https://github.com/feross/webtorrent/blob/master/lib/torrent.js#L235
```js
    parseTorrent.remote(torrentId, function (err, parsedTorrent) {
      if (self.destroyed) return
      if (err) return self._destroy(err)
      self._onParsedTorrent(parsedTorrent)
    })
```

So this method should do a `dht.get(target)`. Problem is that I'd need to pass the current `dht` instance and `parseTorrent` doesn't have that API. Perhaps `parse-torrent` could be an EventEmitter so we don't need to differentiate between .remote() and sync requests.

PS: I hope this level of detail is ok to discuss here. As I said, I'm quite a noob with code, but I'd like to nail down some details before submitting a PR.

**Comment by lmatteis** [comment]
Ok so I've added a pull request with the syntax `parseTorrent.remote({ torrentId: ..., dht: ... }, function () { ... })` and also some tests: https://github.com/feross/parse-torrent/pull/38

Travis tests seem to break because I'm using `ed25519-supercop` for the tests, and it seems incompatible with old node versions?

**Comment by yciabaud** [comment]
I believe this is more likely a travis env issue. 

The default compiler must be to old to handle C++11 like said in: https://travis-ci.org/feross/parse-torrent/builds/153981472#L165

To make it work in this repo we need to add the g++ package to the travis image like @substack did : https://github.com/substack/ed25519-supercop/blob/master/.travis.yml#L10

**Comment by lmatteis** [comment]
@yciabaud I'm really bad with travis (in fact never used it before). What line should I edit exactly to make it work? Feel free submitting a PR on my fork.

**Comment by Jaman1993** [comment]
Any news on this?

**Comment by lmatteis** [comment]
To get things going again @feross there's also this PR on bittorrent-dht waiting review https://github.com/webtorrent/bittorrent-dht/pull/136

**Comment by alanhoff** [comment]
Any updates? :-)

**Comment by innovator256** [comment]
Seriously guys, any updates on this ?

**Comment by HostFat** [comment]
This will be very useful :)

**Comment by Ferk** [comment]
> I was thinking that most of the time publishers will apply updates to the underlying data their sharing. So re-downloading the whole data would be bad. For instance, imagine Archive.org publishing data dumps. As a user I wouldn't want to delete and replace the whole dump every time.

I've been thinking for a while that the ideal approach might be for it to be possible to continue being connected to the old swarm for the chunks that are known to hold the same unchanged data.

This way the swarm would not divide its efforts. If the publisher decides to make changes that you do not want to keep, or for whatever reason you want to be able to access an old version of the torrent, you would be able to do so without splitting the swarm in two.

This would also allow for someone to make a "fork" of any torrent he wants and add/change some data to improve the content without having to start from scratch with a new swarm and new peers, which would split the community and discourage anyone willing to improve the content of a torrent.

It would be great if there was a BEP that allowed something similar as to how GIT works, where you can make a commit on top of another yet keep that commit pointing to the old one allowing you to even decide to "rebase" it later on.

The autoupdate BEP46 brings is convenient, but I would not mind that much updating manually and having the control to decide when do I want to update. The problem with torrents is not really that people don't update, the problem is that when people who don't update they do not contribute to the up-to-date swarm and it loses peers/seeders.

**Comment by RangerMauve** [comment]
One use case that really excites me about this is the ability to load static websites for a browser.

Sort of like what IPFS is doing, but using existing technology instead of building a new tech stack.

Since webtorrent is being integrated with Brave, it's a perfect avenue for building fully decentralized web apps without needing to invest in any specific cryptocurrency.

**Comment by RangerMauve** [comment]
I recently published a library, [mutable-webtorrent](https://github.com/RangerMauve/mutable-webtorrent) which wraps over the WebTorrent API and adds support for mutable torrents in magnet links as well as some helper functions for creating and updating mutable torrents.

I want to revive this effort if possible and get my changes merged into the main webtorrent repo.

**Comment by resession** [comment]
Yeah adding bep 46 support would be great. Can you add the changes @RangerMauve made?

**Comment by anesuc** [comment]
I am commenting merely to get notifications from this fantastic thread haha

**Comment by frafra** [comment]
> I am commenting merely to get notifications from this fantastic thread haha

There is a subscription button on the right for that, no need to write comments.

**Comment by anesuc** [comment]
@frafra Thought there had to be one but didn't see it haha

**Comment by resession** [comment]
any new developments on this?

**Comment by polarathene** [comment]
> any new developments on this?

If there ever is, you'll find it on this issue... just :+1: the top message and click "Subscribe" button.

No need to ask redundantly, just look at what was said before you. Your comment pings everyone who is subscribed for updates, so please refrain from this.

If you'd like to see more traction on it, consider setting up a bounty to fund an incentive for someone to spend time on it.

**Comment by polarathene** [comment]
I do have some questions myself.

The discussion earlier was about using DHT for this functionality, I was under the impression DHT wasn't used with WebTorrent due to WebRTC limitation or something along those lines? Is the proposal only useful for non-WebTorrents?

I would like to implement as a content host for user published content a way to leverage BitTorrent for easing bandwidth expenses, and it seems that'd work well for popular content, but seeding would be less beneficial when publishers release new updates for reasons @Ferk addressed.

The content we host is often data for third-party software to integrate with, so the idea of updating a torrent in place that keeps any prior file state that no longer applies to the latest version would be bad and confuse users with potential breakage experienced. It'd seem better to use separate locations on disk, or replace the previous contents after receiving the new version contents.

AFAIK, presently Broadcatching (adding torrents automatically by monitoring an RSS feed) is a common existing alternative to allow users to receive the latest content and keep seeding it. For some here, that might be just as good of a solution as a mutable torrent, albeit perhaps not as good UX for users?

Regarding the disjointed swarms between updates for seeding, I'm not sure but I think from what [I've read about BitTorrent v2](https://blog.libtorrent.org/2020/09/bittorrent-v2/) (which isn't widely adopted yet), individual files are hashed and can cross torrent boundaries in a swarm when that hash for the same file exists in another torrent.

If so, it'd be interesting to know what happens when you are seeding torrents with existing file content and add a new one, that perhaps the torrent client would pull from those existing local files instead of depending on peers to retrieve it. Regardless it should help work around that problem of seeding an outdated version of content somewhat if some of the content is identical.

**Comment by resession** [comment]
> > any new developments on this?
> 
> If there ever is, you'll find it on this issue... just 👍 the top message and click "Subscribe" button.
> 
> No need to ask redundantly, just look at what was said before you. Your comment pings everyone who is subscribed for updates, so please refrain from this.
> 
> If you'd like to see more traction on it, consider setting up a bounty to fund an incentive for someone to spend time on it.

How do we setup up bounties?

**Comment by anesuc** [comment]
> > any new developments on this?
> 
> If there ever is, you'll find it on this issue... just  the top message and click "Subscribe" button.
> 
> No need to ask redundantly, just look at what was said before you. Your comment pings everyone who is subscribed for updates, so please refrain from this.
> 
> If you'd like to see more traction on it, consider setting up a bounty to fund an incentive for someone to spend time on it.

I don't think he commented to be informed. He genuinely wanted to bump the thread. Sometimes issues like this die and only after someone bumps does someone realise they might have a solution to the problem (they later obtained and just never came back to this thread). I have witnessed this a few times and have myself resolved an issue only after someone bumped it and I actually had a solution I had locally. Your comment does come across as a little aggressive which doesn't really help outside just being aggressive.

**Comment by polarathene** [comment]
> How do we setup up bounties?

Using a third-party service, here's a few:

- [Gitcoin](https://gitcoin.co/bounties/funder)
- [BountySource](https://www.bountysource.com/)
- [GitPay](https://gitpay.me/#/)

**Comment by polarathene** [comment]
> He genuinely wanted to bump the thread. Sometimes issues like this die and only after someone bumps does someone realise they might have a solution to the problem

That's a fair point. Although it's uncommon in my experience I have seen it a few times.

> Your comment does come across as a little aggressive which doesn't really help outside just being aggressive. It might not have been your intention though.

It was unintentional. I just get lots of notifications lately which turn out to be +1 messages or similar, and some of those issues it really is a long string of messages like that, with shorter durations inbetween. 

So apologies to @resession if I came off as too harsh, it had been a good 6 months since any activity in the issue.

**Comment by resession** [comment]
thought of an idea, what if we implement BEP 46/mutability in another package all together. webtorrent can pass in it's instance of the dht and then still have BEP 46.

just testing it for now, i'll post how it goes on the project repo
https://github.com/resession/property

**Comment by rysiekpl** [comment]
Hey all. Any movement on this? Any chance of having BEP46 implemented in `webtorrent`? I am looking at different options for a transport layer of a [thing I'm building here](https://resilient.is/), and mutable torrents [could be one of the options](https://gitlab.com/rysiekpl/libresilient/-/issues/44).

Thanks!

**Comment by RangerMauve** [comment]
We've implemented some stuff for it in bt-fetch https://github.com/RangerMauve/bt-fetch

It's still kinda rough and mutable torrents cannot work on the web since there's no access to the DHT.

**Comment by rysiekpl** [comment]
> We've implemented some stuff for it in bt-fetch https://github.com/RangerMauve/bt-fetch

Oh nice, thanks!

> It's still kinda rough and mutable torrents cannot work on the web since there's no access to the DHT.

Ah right. That's not great, then Good to know though!

**Comment by rysiekpl** [comment]
> If you'd like to get mutable torrents working on the web, I'd love to collaborate on creating a new bittorrent extension type that I've been ideating that would work similar to the pubsub stuff described in BEP 50: http://bittorrent.org/beps/bep_0050.html

The reason I am interested of mutable torrents is because of a project I am working on, called [LibResilient](https://resilient.is/); the idea is to use Service Workers in browsers to make websites still work even if the main domain is down. I a looking at WebTorrents as one possible transport ([issue here](https://gitlab.com/rysiekpl/libresilient/-/issues/44)).

So I am more interested in stuff that already exists and *works in the browser* (specifically in the Service Worker context).

**Comment by RangerMauve** [comment]
From what I recall, WebRTC doesn't work in ServiceWorkers so you're pretty much stuck with using servers + the regular offline cache. 😅 WebTorrent probably wouldn't add anything useful that you couldn't already do with basic service worker tech.

**Comment by lemanschik** [comment]
@RangerMauve it is even more bad then that the only thing i am worrying about is that the code is npm based splitted over tons of repos i am not sure if i can reuse any of this code or if i need to rewrite it or

the most prominent idea at present if no one has a better is that i release simply bittorrent for js? and we call web torrent dead?

**Comment by ThaUnknown** [comment]
@lemanschik there's a very valid reason the code is split like that

you can think of it as easily swappable modules, you can implement API's for those modules in a different package and just swap them if you want to change something, pretty much all web API's are built like that

webtorrent is far from dead, tho a lot closed than it used to be in past times, the main reason for this is that webtorrent is community driven, and not many people are interested in BT in web, and unlike say qbit or libtorrent, wt doesn't make any money from this and doesn't have a following

if you want BEP46 simply code it in as a module, PR it, it gets reviewed, and added, this is what I've been doing with webtorrent for a few years now, and I've definitely done my part in making webtorrent better even with my somewhat limited sets of skills and it honestly makes me feel sad that noone wants to contribute

**Comment by lemanschik** [comment]
@ThaUnknown i guess your not fully getting the picture but that does not matter for your interrests i am chromium engineer and i have implemented the direct sockets api 

https://chromestatus.com/feature/6398297361088512

but i did not update the status but i can use it so you can imagine what that means i can replace Web Torrent with a transport agnostic implementation directly running in browsers today even connections RAW TCP UDP are not blocking me from doing so

the main miss assumption from you was that i am a end user like you but i am Browser Architectural Platform Lead Engineer so i have a total other view then you on this project i in fact will end this project and replace it with nativ in browser methods and you can think and say what you like about the future of this.

hope that makes sense and answered the questions. as also i am writing whole essays about switching away from oldschool package management nodejs and all this ...........

- https://github.com/lemanschik/why-you-should-switch-away-from-npm-nodejs/blob/main/README.md

**Comment by ThaUnknown** [comment]
okay well I just got fed up so I'm shooting as straight as possible:

@lemanschik honestly I couldn't care less?

you're a chromium dev which implemented stuff unrelated to webtorrent, which webtorrent doesn't use.... what do you expect me to say here? good for you?

you know VERY well that TCP/UDP sockets in chromium will never happen because of the security concerns, and I'm fairly sure a proposal was already denied once, and even if they will be implemented, you know very well that webtorrent's peer's will simply have another dependency to use chromium TPC/UDP sockets aka another npm package ;^)

I understand you're probably well capable of "replacing webtorrent with a browser API" but not only will that not be implemented into chromium, I don't see the need of you just flailing that flag around? you're going I CAN MAKE YOU USELESS IF I WANTED TO.... great?? go do it then?

the fact that you dislike node modules and don't want them used is.... I mean it has basis, and you're probably not alone, but wanting something, and it actually being accepted and used are 2 different stories, shit look at ESM, in theory it's fantastic and has only upsides from say nodeJS require, yet people don't want to use it because it requires change and a perfect example of that is `node-fetch`

I'm aware you're probably well capable of helping webtorrent and or improving it, but as of now you're just flailing that flag and I don't really know... expecting people to go "please help us our savior"?, if you like this package, contribute, make a PR, it will get reviewed pretty earnestly [for FOSS], you'll get a nice "thank you" a pat on the back and that's about it, since again this is FOSS, and not the chromium dev you're used to where people expect stuff actually being done

pretty much nothing you said is useful in regards to implementing BEP46 and I'm honestly struggling to bring out any value in what you're trying to convey, and trust me I'm trying

**Comment by lemanschik** [comment]
@ThaUnknown ok i expand exactly on Mutable torrents they are a thing of the past the new web works on other principles distributed net wide constant loaded Memory that means you can import also running applications. Which Means you can import a torrent and it can emit dynamic data maybe that solves the missing parts about your BEP46 it is not the torrent protocol it self. but again the original torrent protocol is not needed 

and all torrent clients can be compiled to wasm and use the tcp udp sockets they work out of the box.

**Comment by lemanschik** [comment]
It looks more like that 
```ts
// ref is org/user/${githubref}
const { resolvedId } = import('https://github.lemanschik.com/resolve/resolve.js?github=ref');
const .........
```

but translated to torrent your importing it via a ESModule and the torrent hash the rest gets solved dynamic already

**Comment by ThaUnknown** [comment]
sure?...
with hindsight BEP46 implementation has it's issues but this isn't about making a new decentralized protocol stack, this is about implementing existing specs, what you're suggesting is changing a spec established as well as IRC to your likings and expecting people to follow suit, and even if all torrent clients implemented it RIGHT NOW, probably <15% of all peers would actually use that implementations because lets face it, people, seedboxes and private trackers rarely update their clients

I don't understand the WASM stab again, it seems very irrelevant to webtorrent, and it's already been proven that WASM rarely actually makes shit faster when it has to interact with web API's anyways

again I don't understand the point of you showing an example async import from an http remote, it adds extra error points and makes bundling more annoying, a perfect example is that github imports don't work on IPv6 machines

**Comment by lemanschik** [comment]
@ThaUnknown i do not know why it does not work for you exactly or where you see problems i agree that the http example looks confusing in the first place because i did not expand on the full protocol

the resolve part returns a resolvedId as permernentLoaded String in the engine that can get reused internal 
I need to hire people to do graphics as that gets complicated the most easy to explain it is when i abstract everything into 2 Types Streams and Data and i do not go into what data is

You will use the resolvedId to connect to a Stream that will send Data Dynamic one it will contain again memory pointers represented as Strings and this way Compose a memory Table like a hashmap that it can directly access via Handels.

The handels which are abstract object do contain logic for security and failover they are able to replace them self for example if the target peer goes offline and all this situations 

that full algo can get complex. But we use that algo all over internal i only make it more useful for users directly.

at the end the only thing that is importent out of high level view is you will not fetch a webtorrent file and open that via a webtorrent client. You will import a common ESModule that can represent a torrent connected resource that you then get as stream that you can write to a file or pipeThroug a network and so on

**Comment by ThaUnknown** [comment]
@RangerMauve I just had a look at your packages [sorry it's taking so long, but I'm out here trying to keep webtorrent alive solo], how usable are they, would you be interested in adding bep46 to webtorrent? web DHT isn't here yet, and idk when that will happen but this could at the very least help node

**Comment by RangerMauve** [comment]
they are decently usable in my experience. it can be a little sketchy when the dht is having a rough time.  I also think we are missing some details in marking torrents as mutable when they are created which I think is handled in libtorrent.

I am currently using my implementations and the distributed press project and in Agregore.  hopefully we will see more stats as we increase the number of users.

I don't know how much time I would have to devote on this, but I would be down to have a call or something to scope out the approach.  if anything, I think we could add in the code from my [mutable web torrent](https://github.com/RangerMauve/mutable-webtorrent)  library and go from there.

I agree that this would only make sense for node, however I have some thoughts on how we can get the web to support these things you without needing the distributed hash table.

**Comment by lemanschik** [comment]
@RangerMauve wait some more weeks Isolated Web APPS got pushed already this includes Direct Sockets see.https://github.com/GoogleChromeLabs/telnet-client

**Comment by ThaUnknown** [comment]
@lemanschik i said this before once, and i dont really like repeating myself, this isn't about alternative solutions, we want that functionality in webrtc, so please stop derailing conversations

@RangerMauve thanks for the reply, if you use discord you can just join webtorrent's discord and ping me there, that said I'm not certain when I'll have time so we'd have to schedule a meet in DMs 

the main concerns is compatibility with existing torrent clients for nodejs, and structure that wouldnt a big rewrite for web and ofc tests

**Comment by lemanschik** [comment]
@ThaUnknown it is in WebRTC already you only need to write the instructions? You do not get the whole context. The Problem is You can not Handle a Browser like a Normal Server because!!!!!

Then the Browser would be Target to the Same ISSUES as a Server. That means a lot of more Security Monitoring Pro Active Network Management and so on.

The Browser has a Alternative way to archive the same goals what ever you want does not matter as the browser is a platform used by millions we need to take care that millions are happy with it and that it does not introduce more problems then it solves you maybe agree with that.

Web Torrent in the Modern world is simple It can be implemented in a few lines of JS Code

but you need to strip down direct TCP UDP Connections in favor of user Accepted TLS Secured UDP Sessions that's it. 

The User needs some code that queries what it should accept on its data channels and what not.

So at present the current state is to define a new BEP that allows all people to control and manage shares. 

The Current Solutions did not define that they say use simple as much Resources as available.

## What parameters are needed
At present we define so called DHT and or TorrentTrackers
this need to get replaced by something that is http based so we can reach it with text protocols. Also the Information that is tracked changes a lot

eg: before we tracked hash, ip , port combos today we need to track SDP Strings. 

There are up to a 1000 small pieces that need adjustment and i am working on that since over 10 years. 

The Demo above is by the Way Fully in the browser and when you compile libtorrent to wasm you can use it in the telnet app example without problems but your not able to share via TCP but you can use the TCP Trackers. and you can use UDP with nativ lib torrent clients.  

this is because of TCPOpen is only able to do outside connections from you to a target not from a target to you 

so the next step to get the old torrent stuff working is to get a way and protocol to allow incoming connections via WebRTC.

My current concept is to use a https origin and then The Origin Server it Self is WebRTC + Quic + (Web 4.0 Internal Iteration over a new Ethernet Protocol) Enabled. Such a Origin Server is able to share SDP so the WebRTC Connection infos via HTTP and you can then use that as Tracker

```

const WebRTCEnabledOriginServer = {}
const PeerConnectionWithTorrentNetwork = fetch(WebRTCEnabledOriginServer);
const index = PeerConnectionWithTorrentNetwork.createDataChannel('DHT')
// WebRTCEnabledOriginServer and this client do use whats called Perfect Negotiation
// That means simple sayed they can switch host and server positions at any time.
index.postMessage("GET HASH");
index.once('message', (msg) => 
/** response from get hash is a datachannel hash 
 * you can now iterate over your datachannels to get the download or do a upload.
 */)
```

The unsolved part is we need to emit GET Hash on clients that do contain chunks of that TorrentHash. to let them join the hash datachannel on the Tracker and get a own datachannel session for input or output the direction does not matter.
